src/Forum/ValidateMailConfiguration.php

```diff
@@ -0,0 +1,89 @@
+<?php
+
+/*
+ * This file is part of Flarum.
+ *
+ * (c) Toby Zerner <toby.zerner@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Flarum\Forum;
+
+use Flarum\Foundation\Application;
+use Flarum\Mail\DriverInterface;
+use Flarum\Mail\NullDriver;
+use Flarum\Settings\Event\Saving;
+use Flarum\Settings\SettingsRepositoryInterface;
+use Flarum\Settings\SettingsServiceProvider;
+use Illuminate\Contracts\Container\Container;
+use Illuminate\Support\Arr;
+use Illuminate\Validation\Validator;
+
+class ValidateMailConfiguration
+{
+    /**
+     * @var SettingsServiceProvider
+     */
+    protected $settings;
+
+    /**
+     * @var Application
+     */
+    protected $container;
+
+    /**
+     * @param Container $container
+     * @param SettingsRepositoryInterface $settings
+     */
+    public function __construct(Container $container, SettingsRepositoryInterface $settings)
+    {
+        $this->container = $container;
+        $this->settings = $settings;
+    }
+
+    public function whenSettingsSaving(Saving $event)
+    {
+        if (! isset($event->settings['mail_driver'])) {
+            return;
+        }
+
+        $driver = $this->getDriver($event->settings);
+
+        $this->getValidator($driver, $event->settings)->validate();
+    }
+
+    public function getWorkingDriver()
+    {
+        $settings = $this->settings->all();
+        $driver = $this->getDriver($settings);
+        $validator = $this->getValidator($driver, $settings);
+
+        return$validator->passes()
```

Comment:
```suggestion
        return $validator->passes()
```

How .. is this even working 🙈

Also `getWorkingDriver` does not implicate what this method exactly does. I think it returns the configured driver if validated, otherwise return the fallback. Personally I think this could all be part of `getDriver` instead, but not sure whether that method would explode..

Reply:
Not sure why it worked nor why StyleCI did not check my commits 🤔 

`getDriver` returns an instance of the driver set through the settings if it exists, else it returns the `NullDriver`

`getWorkingDriver` checks for validation of that configured driver and returns it if the validation passes, else it returns the `NullDriver`.

I didn't like the implementation either.

Reply:
To be honest, the purpose of the whole class is a bit unclear to me. It seems to be both a driver factory and a validator of some sort, that is instantiated when creating drivers (e.g. for sending) and when storing settings.

I guess I could see more clearly if the validation logic would go to each driver implementation. Essentially a method that you can ask whether the current configuration is valid, and which exact problems exist if it isn't.